### ISSUE TITLE
ios_config with src parameter not functioning correctly with ACLs

### ISSUE THREAD

**Issue by bdlamprecht** [ansible/ansible#25449]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
ios_config

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/my_modules/']
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
Standard

##### OS / ENVIRONMENT
Ubuntu 16.04.2 LTS

##### STEPS TO REPRODUCE
When running the PB below and making changes to the ACL contained within the referenced `src` file (i.e. udp port destination numbers), if there are differences, it simply adds a new entry to the bottom of the ACL instead of replacing it in the correct location and order of the ACL.

Playbook:
```
  - name: Verify IP SLA ACLs Are Correct
    ios_config:
      authorize: yes
      provider: "{{ creds }}"
      src: "{{ role_path }}/templates/ipsla/{{ platform }}/ip_access-list_ipsla.cfg"
      match: exact
      replace: block
```

Contents of `src` file referenced above:
```
ip access-list extended al_att_ipsla_class1
 remark /**********************************************
 remark QoS class 1 - classification policy - IP SLA
 remark **********************************************/
 permit udp any any range 17100 17199
 permit udp any range 17100 17199 any
```

##### EXPECTED RESULTS
If I were to change, for example, in the above ACL the udp destination port range from `17100 17199` to `17100 18000` and re-run the PB, it would remove the ACL, and re-create it properly.

##### ACTUAL RESULTS
The new and modified ACL entries in the `src` file are simply placed at the end of the ACL on the device.
The following is a run-through of how this was tested.

---The ACL is manually removed---
```
SWITCH#conf t
Enter configuration commands, one per line.  End with CNTL/Z.
SWITCH(config)#no ip access-list extended al_att_ipsla_class1
SWITCH(config)#^Z
```
---PB is executed with the original ACL---
```
SWITCH#sh run | s ip access-list extended al_att_ipsla_class1
ip access-list extended al_att_ipsla_class1
 remark /**********************************************
 remark QoS class 1 - classification policy - IP SLA
 remark **********************************************/
 permit udp any any range 17100 17199
 permit udp any range 17100 17199 any
```
--- So far, everything is correct ---
--- UPDATED `src` file (note the different ending ports)---
```
ip access-list extended al_att_ipsla_class1
 remark /**********************************************
 remark QoS class 1 - classification policy - IP SLA
 remark **********************************************/
 permit udp any any range 17100 18000
 permit udp any range 17100 18000 any
```
---PB is re-run with UPDATED `src` file---
```
SWITCH#sh run | s ip access-list extended al_att_ipsla_class1
ip access-list extended al_att_ipsla_class1
 remark /**********************************************
 remark QoS class 1 - classification policy - IP SLA
 remark **********************************************/
 permit udp any any range 17100 17199
 permit udp any range 17100 17199 any
 permit udp any any range 17100 18000
 permit udp any range 17100 18000 any
```
As you can see, the ACL is updated with the new entries, but the old ones are not removed.

If this is not the intended usage, I would at least like to request that the documentation for `ios_config` to reflect this confusion.

**Comment by ansibot** [comment]
cc @privateip @rcarrillocruz
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by privateip** [comment]
@bdlamprecht The `ios_config` module has an argument to remove the ACL before re-applying it.  The module doesn't implement heuristics to try to update an ACL line.  

I need to test this but this serve as a workaround for the moment.

```
  - name: Verify IP SLA ACLs Are Correct
    ios_config:
      authorize: yes
      provider: "{{ creds }}"
      lines: "{{ lookup('file', '{{ role_path }}/templates/ipsla/{{ platform }}/ip_access-list_ipsla.cfg' }}.split("\n')"
      match: exact
      replace: block
      before: no ip access-list extended al_att_ipsla_class1
```

**Comment by bdlamprecht** [comment]
Besides some minor typos in your example above, that `task` does work, however, it appears to re-apply the ACL every time. 

What I'm trying to accomplish is to check if the `ACL` matches the supplied `.cfg` file and one of two outcomes should take place:
1. If it does, nothing needs to be done and no alterations should be performed.
2. If not, the `ACL` should be removed and then re-applied.

Perhaps, I'm going about this with an entirely wrong approach, but would appreciate any help or advice you can provide.

Thanks.

**Comment by privateip** [comment]
@bdlamprecht The following should accomplish what you are trying to do:

```
  - name: Verify IP SLA ACLs Are Correct
    ios_config:
      authorize: yes
      provider: "{{ creds }}"
      lines: "{{ lookup('file', '{{ role_path }}/templates/ipsla/{{ platform }}/ip_access-list_ipsla.cfg' }}.split("\n')"
      match: exact
      replace: block
      before: no ip access-list extended al_att_ipsla_class1
      parents: ip access-list extended al_att_ipsla_class1
```

In the task above, notice the addition of the parents argument that is set essentially the top level ```ip access-list ...``` command.  The entries in the lookup file then should _only_ be the acl entries.  For example the contents of ```ip_access-list_ipsla.cfg``` might be:

```
remark /**********************************************
remark QoS class 1 - classification policy - IP SLA
remark **********************************************/
permit udp any any range 17100 17199
permit udp any range 17100 17199 any
```

When using the ```match: exact``` argument, the contents of the file must reflect *exactly* how the entries appear in the running config including order otherwise it will trigger a change.

**Comment by bdlamprecht** [comment]
@privateip I was about to write a long comment providing evidence where the addition of the `parents` parameter didn't change anything in the outcome of the `PLAY RECAP`, but right before submitting that comment, I realized in your statement above concerning the `cfg` file, you hadn't included the `ip access-list extended al_att_ipsla_class1 ` as the first line.

When I tried that, it *__DID__* solve the problem I was encountering.  I'm happy to use this solution and I appreciate you helping me with it.

That being said, it appears to be somewhat of a hack.  I'm not sure of what else can be done, but perhaps in a future release of Ansible, there could be a `module` for updating and validating ACLs on IOS devices.  This may be a long short with the lack of vendor support on a platform that doesn't have any APIs to speak of.  Obviously this is a wish list of my own and I would love to help to script it, however, I don't currently have the Python skills to do so.

In closing, thanks a lot. :smile:
